File: csharp/src/EMAYBLEClient.cs
/*
 * EMAY SleepO2 — C# types, protocol, and BLE client
 * Copyright (c) 2026 Ground Effect Software, LLC · MIT
 *
 * BLE backends:
 *   Windows:  Windows.Devices.Bluetooth (UWP / .NET 8+)
 *   Linux:    BlueZ via D-Bus (Tmds.DBus or similar)
 *   macOS:    CoreBluetooth via Xamarin.Mac or .NET MAUI
 *   Android:  Android.Bluetooth via Xamarin.Android / .NET MAUI
 *
 * This reference implementation shows the abstract workflow.
 * A production library would use a platform DI strategy.
 */
namespace GroundEffectSoftware.EMAYSleepO2;

using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

/* ---- Types ---- */

public record EMAYReading(int? Spo2, int? Pulse, long TimestampMs);

public record MinuteSample(long MinuteStartMs, string MetricType, double Value, string UnitString);

/* ---- Protocol ---- */

public static class EMAYProtocol
{
    public static readonly Guid SvcUuid = Guid.Parse("0000ff12-0000-1000-8000-00805f9b34fb");
    public static readonly Guid WrUuid  = Guid.Parse("0000ff01-0000-1000-8000-00805f9b34fb");
    public static readonly Guid NfyUuid = Guid.Parse("0000ff02-0000-1000-8000-00805f9b34fb");
    public const string NamePrefix = "SleepO2";

    public static readonly byte[] Hello       = { 0x89, 0x09 };
    public static readonly byte[] DeviceState = { 0x8E, 0x05, 0x13 };
    public static readonly byte[] StartCmd    = { 0x9B, 0x01, 0x1C };
    public static readonly byte[] StopCmd     = { 0x9B, 0x7F, 0x1A };
    public static readonly byte[] Battery     = { 0x86, 0x06 };
    public static readonly byte[] Heartbeat   = { 0x9A, 0x1A };
    public static readonly byte[][] StartSeq  = { Hello, DeviceState, StartCmd, Battery };

    public static byte Checksum(byte[] p) => (byte)(p.Sum(b => (int)b) & 0x7F);
    public static byte[] Command(byte[] p) => [..p, Checksum(p)];

    public static EMAYReading? Parse(byte[] raw)
    {
        if (raw is not { Length: 8 }) return null;
        if (raw[0] != 0xEB || raw[1] != 1 || raw[2] != 5) return null;
        if (raw[5] != 0x7F || raw[6] != 0) return null;
        int sum = 0; for (int i = 0; i < 7; i++) sum += raw[i];
        if (raw[7] != (byte)(sum & 0x7F)) return null;
        int pr = raw[3], so = raw[4];
        int? pulse = (pr == 0 || pr == 0xFF) ? null : pr;
        int? spo2  = (so == 0 || so == 0xFF) ? null : so;
        if (pulse is < 30 or > 220) return null;
        if (spo2 is < 0 or > 100) return null;
        return new EMAYReading(spo2, pulse, DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
    }
}

/* ---- BLE Client (abstract — backends below) ---- */

public abstract class EMAYBLEClient : IDisposable
{
    public event Action<EMAYReading>? OnReading;
    public event Action<string>? OnStatus;

    protected void Emit(EMAYReading r) => OnReading?.Invoke(r);
    protected void Emit(string s) => OnStatus?.Invoke(s);

    public abstract Task StartAsync(CancellationToken ct = default);
    public abstract Task StopAsync();

    public virtual void Dispose() => StopAsync().GetAwaiter().GetResult();
}

/* ---- Windows BLE backend ---- */
#if WINDOWS

using Windows.Devices.Bluetooth;
using Windows.Devices.Bluetooth.Advertisement;
using Windows.Devices.Bluetooth.GenericAttributeProfile;
using Windows.Storage.Streams;

public class EMAYWindowsBLEClient : EMAYBLEClient
{
    private BluetoothLEDevice? _device;
    private GattCharacteristic? _wr, _nfy;
    private Timer? _heartbeat;

    public override async Task StartAsync(CancellationToken ct = default)
    {
        Emit("scanning");
        var watcher = new BluetoothLEAdvertisementWatcher
        {
            ScanningMode = BluetoothLEScanningMode.Active
        };
        var tcs = new TaskCompletionSource<BluetoothLEDevice>();

        watcher.Received += (s, args) =>
        {
            if (args.Advertisement.LocalName?.StartsWith(EMAYProtocol.NamePrefix) == true)
            {
                watcher.Stop();
                _ = Task.Run(async () =>
                {
                    var dev = await BluetoothLEDevice.FromBluetoothAddressAsync(args.BluetoothAddress);
                    if (dev != null) tcs.TrySetResult(dev);
                }, ct);
            }
        };
        watcher.Start();
        _device = await tcs.Task;

        Emit("connecting");
        var svcResult = await _device.GetGattServiceAsync(EMAYProtocol.SvcUuid);
        var svc = svcResult.Service;
        _wr  = (await svc.GetCharacteristicsAsync(EMAYProtocol.WrUuid)).Characteristics.FirstOrDefault();
        _nfy = (await svc.GetCharacteristicsAsync(EMAYProtocol.NfyUuid)).Characteristics.FirstOrDefault();

        await _nfy!.WriteClientCharacteristicConfigurationDescriptorAsync(
            GattClientCharacteristicConfigurationDescriptorValue.Notify);

        Emit("streaming");
        _nfy.ValueChanged += (s, args) =>
        {
            var r = EMAYProtocol.Parse(args.CharacteristicValue.ToArray());
            if (r != null) Emit(r);
        };

        foreach (var cmd in EMAYProtocol.StartSeq)
            await _wr.WriteValueAsync(cmd.AsBuffer());

        _heartbeat = new Timer(async _ =>
        {
            if (_wr != null)
                await _wr.WriteValueAsync(EMAYProtocol.Heartbeat.AsBuffer());
        }, null, 1000, 1000);
    }

    public override async Task StopAsync()
    {
        _heartbeat?.Dispose();
        _device?.Dispose();
        Emit("stopped");
        await Task.CompletedTask;
    }
}
#endif

/* ---- Linux BlueZ backend (reference — requires Tmds.DBus) ---- */
#if LINUX
public class EMAYLinuxBLEClient : EMAYBLEClient
{
    // Uses org.bluez via Tmds.DBus.
    // Workflow: Adapter1.StartDiscovery → Device1.Connect →
    //           GattCharacteristic1.StartNotify → WriteValue for commands.
    // Full implementation ~150 lines of D-Bus method calls.
    public override Task StartAsync(CancellationToken ct = default)
    {
        Emit("Linux BlueZ BLE client — requires Tmds.DBus and org.bluez");
        return Task.CompletedTask;
    }
    public override Task StopAsync() { Emit("stopped"); return Task.CompletedTask; }
}
#endif

File: php/src/EMAYBLEClient.php
<?php
/*
 * EMAY SleepO2 — PHP types, protocol, and BlueZ FFI BLE client
 * Copyright (c) 2026 Ground Effect Software, LLC · MIT
 *
 * PHP is not a native BLE runtime. This reference implementation
 * provides the protocol layer plus an FFI bridge to libbluetooth
 * for Linux/BlueZ. On other platforms, use a sidecar process
 * (Python/C) for BLE I/O and communicate via pipe/socket.
 *
 * Requires PHP 8.1+ with FFI enabled, and libbluetooth:
 *   sudo apt install libbluetooth-dev
 *   echo 'ffi.enable=true' >> /etc/php/8.3/cli/conf.d/20-ffi.ini
 */

declare(strict_types=1);

namespace GroundEffectSoftware\EMAYSleepO2;

/* ---- Types ---- */

class EMAYReading
{
    public function __construct(
        public readonly ?int $spo2,
        public readonly ?int $pulse,
        public readonly float $timestampSecs,
    ) {}
}

class MinuteSample
{
    public function __construct(
        public readonly float $minuteStartSecs,
        public readonly string $metricType,
        public readonly float $value,
        public readonly string $unitString,
    ) {}
}

/* ---- Protocol ---- */

final class EMAYProtocol
{
    public const string SVC_UUID = '0000ff12-0000-1000-8000-00805f9b34fb';
    public const string WR_UUID  = '0000ff01-0000-1000-8000-00805f9b34fb';
    public const string NFY_UUID = '0000ff02-0000-1000-8000-00805f9b34fb';
    public const string NAME_PFX = 'SleepO2';

    public const string HELLO        = "\x89\x09";
    public const string DEVICE_STATE = "\x8E\x05\x13";
    public const string START_CMD    = "\x9B\x01\x1C";
    public const string STOP_CMD     = "\x9B\x7F\x1A";
    public const string BATTERY      = "\x86\x06";
    public const string HEARTBEAT    = "\x9A\x1A";

    public const array START_SEQ = [
        self::HELLO, self::DEVICE_STATE, self::START_CMD, self::BATTERY
    ];

    public static function checksum(string $p): int
    {
        return array_sum(array_map('ord', str_split($p))) & 0x7F;
    }

    public static function command(string $p): string
    {
        return $p . chr(self::checksum($p));
    }

    public static function parse(string $raw): ?EMAYReading
    {
        if (strlen($raw) !== 8) return null;
        $b = array_map('ord', str_split($raw));
        if ($b[0] !== 0xEB || $b[1] !== 1 || $b[2] !== 5) return null;
        if ($b[5] !== 0x7F || $b[6] !== 0) return null;
        if ($b[7] !== (array_sum(array_slice($b, 0, 7)) & 0x7F)) return null;
        $pulse = in_array($b[3], [0, 0xFF], true) ? null : $b[3];
        $spo2  = in_array($b[4], [0, 0xFF], true) ? null : $b[4];
        if ($pulse !== null && ($pulse < 30 || $pulse > 220)) return null;
        if ($spo2  !== null && ($spo2  < 0  || $spo2  > 100)) return null;
        return new EMAYReading($spo2, $pulse, microtime(true));
    }
}

/* ---- FFI BLE Client (Linux/BlueZ) ---- */

final class EMAYBLEClient
{
    private ?\FFI $ffi = null;
    private ?\Closure $onReading = null;
    private ?\Closure $onStatus = null;
    private bool $running = false;

    /**
     * @param callable(EMAYReading):void $onReading
     * @param callable(string):void $onStatus
     */
    public function __construct(
        ?callable $onReading = null,
        ?callable $onStatus = null,
    ) {
        $this->onReading = $onReading ? $onReading(...) : null;
        $this->onStatus  = $onStatus  ? $onStatus(...)  : null;

        // FFI bridge to libbluetooth (HCI sockets)
        // HCI socket operations for BLE scanning are low-level;
        // a production implementation would use a sidecar or
        // the C reference client via FFI.
        try {
            $this->ffi = \FFI::cdef('
                typedef struct { int dd; } hci_dev;
                int hci_open_dev(int dev_id);
                int hci_close_dev(int dd);
                int hci_le_set_scan_parameters(int dd, int type, int interval, int window,
                                                int own_type, int filter_policy, int to);
                int hci_le_set_scan_enable(int dd, int enable, int filter_dup, int to);
            ', 'libbluetooth.so');
        } catch (\FFI\Exception) {
            $this->emit('FFI: libbluetooth.so not available — BLE requires a sidecar');
        }
    }

    public function start(): void
    {
        $this->emit('scanning for SleepO2...');
        /*
         * Reference workflow (requires a sidecar or native extension):
         * 1. HCI LE scan for devices advertising the SleepO2 service UUID
         * 2. HCI LE connect
         * 3. GATT discover → find write + notify characteristics
         * 4. Write START_SEQ commands
         * 5. Read notifications → parse frames → call onReading
         * 6. Heartbeat loop every 1 second
         */
    }

    public function stop(): void
    {
        $this->running = false;
        $this->emit('stopped');
    }

    private function emit(string|EMAYReading $event): void
    {
        if ($event instanceof EMAYReading) {
            $this->onReading?->__invoke($event);
        } else {
            $this->onStatus?->__invoke($event);
        }
    }
}
